### PR TITLE
Make it possible to specify a new queue instance

### PR DESCRIPTION
**Pull request by joshdifabio**
In order to enable efficient integrations with reactors, it would be very helpful to be able to provide a custom task queue.

Here's a simple example of what would be possible after merging this PR:

```php
use Amp\Reactor;
use GuzzleHttp\Promise\TaskQueue;

class AmpTaskQueue extends TaskQueue
{
    private $reactor;

    public function __construct(Reactor $reactor)
    {
        $this->reactor = $reactor;
    }

    public function add(callable $task)
    {
        if ($this->isEmpty()) {
            $this->reactor->immediately([$this, 'run']);
        }

        parent::add($task);
    }
}

\GuzzleHttp\Promise\queue(new AmpTaskQueue($reactor));
```

Simply calling `$reactor->immedately($task);` would really be preferable but, unfortunately, this would prevent the `run()` and `isEmpty()` methods from working as expected.

### CODE REVIEW

**Comment by joshdifabio** [comment]
Note that this is the approach `react/promise` is likely to take when they introduce a task queue in v3.

https://github.com/reactphp/promise/blob/master/src/functions.php#L203

**Comment by joshdifabio** [comment]
Any thoughts on this @mtdowling? It would be great for optimisation where the curl handler isn't being used.

**Comment by mtdowling** [comment]
This should be ok considering it would be an application concern to modify this. Do you think it would be better to introduce an interface here?

**Comment by joshdifabio** [comment]
> Do you think it would be better to introduce an interface here?

Probably, yes. Would you like me to add one?

**Comment by mtdowling** [comment]
That would be great!

**Comment by joshdifabio** [comment]
I've added an interface. The main question in my mind was whether `disableShutdown()` belonged in the interface, but for BC I think it probably should.

**Comment by joshdifabio** [comment]
What are your thoughts on this @mtdowling? We've been using this in production for a while and it's been working well for us. Referencing a fork branch in our Composer configs is getting pretty tricky however.

File: src/TaskQueueInterface.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+     * exits IFF the exit is not the result of a PHP E_ERROR error.
+     *
+     * You can disable running the automatic shutdown of the queue by calling
+     * this function. If you disable the task queue shutdown process, then you
+     * MUST either run the task queue (as a result of running your event loop
+     * or manually using the run() method) or wait on each outstanding promise.
+     *
+     * Note: This shutdown will occur before any destructors are triggered.
+     */
+    public function disableShutdown();
```
Comment (mtdowling) [inline src/TaskQueueInterface.php]: I wonder if this should just be an implementation detail of the default `TaskQueue`.

Comment (joshdifabio) [inline src/TaskQueueInterface.php]: I had the same thought. However, would this not be a breaking change? (E.g. `queue()->disableShutdown()` would potentially no longer work.)

Comment (mtdowling) [inline src/TaskQueueInterface.php]: I think anyone calling that method has already coupled themselves to a specific implementation detail though... I would be in favor of removing this from the interface but keeping it on the TaskQueue class.

Comment (joshdifabio) [inline src/TaskQueueInterface.php]: How do you feel about merging this if I update the PR to remove `disableShutdown()` from `QueueInterface`? Requiring this forked version in our code base is getting more and more tricky.

Comment (mtdowling) [inline src/TaskQueueInterface.php]: Yes. Removing this from the interface is the right direction in my opinion. Will be merged after.

**Comment by joshdifabio** [comment]
@mtdowling I've removed `disableShutdown()` from `TaskQueueInterface` as requested.

P.S. Thanks for your responsiveness here and in general, I really appreciate it.

**Comment by mtdowling** [comment]
Thanks! This looks great.

**Pull request merged**